adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%% begin cuda_backend %%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+[appendix]
+[[chapter:cuda-backend]]
+= CUDA backend specification
+
+This chapter describes how the SYCL general programming model is mapped on top
+of CUDA, and how the SYCL generic interoperability interface must be
+implemented by vendors providing SYCL for CUDA implementations to ensure SYCL
+applications written for the CUDA backend are interoperable.
+
+The CUDA backend is enabled using the `sycl::backend::cuda` value of `enum
+class backend`. That means that when the CUDA backend is active, the value of
+`sycl::is_backend_active<sycl::backend::cuda>::value` will be `true`, and the
+preprocessor macro `SYCL_BACKEND_CUDA` will be defined.
+
+The CUDA backend requires an installation of CUDA SDK as well as one or more
+CUDA devices available in the system.
+[[sec:cuda:introduction]]
+== Introduction
+
+[[sec:cuda:mapping_of_sycl_programming_model]]
+== Mapping of SYCL programming model
+
+This section gives a general overview of how the SYCL programming model maps to
+CUDA. These two programming models are pretty similar in essence however they do
+have a few differences in terminology and architecture.
+
+[[sub:cuda:platform_model]]
+=== Platform Model
+
+A SYCL device maps to a single CUDA device.  As CUDA does not split into
+separate platforms there is no 'platform' concept in CUDA corresponding to the
+SYCL platform. Instead, a SYCL platform maps to a collection of CUDA devices
+represented by `std::vector<CUdevice>`.
+
+A SYCL <<context>> simply maps to one, or multiple CUDA contexts. Indeed while
+a CUDA context is tied to a single device, this is not the case for a SYCL
+<<context>> and the CUDA backend implementation may use multiple CUDA contexts
+to emulate a SYCL <<context>> containing multiple devices. Additionally, while
+SYCL contexts are simple objects passed around either implicitly or explicitly,
+CUDA contexts require to be activated on the current thread to be used by other
+CUDA entry points. Therefore any use of the SYCL APIs with a CUDA backend may
+modify the current active context on the thread, and no guarantee is provided
+that any existing active CUDA context would be restored by SYCL.
```

Comment:
cc @npmiller ^

Reply:
So for the context part you are understanding that correctly, and yes it is legal for the application to change the active context before the next SYCL API call. Essentially the SYCL runtime must ensure that it's using the correct context associated with the `sycl::context`, but we don't want to force it to have to restore existing CUDA contexts, as this is a concern for performance and can sometimes lead to errors as well. This stems from issues we had in DPC++ (see write-up here: https://github.com/intel/llvm/pull/4442 ), and trying to keep the guidelines as general as possible so they can work with different implementations.

About the active devices, this is a very good question, so CUDA has two different APIs, the "driver API" which is lower level and used in DPC++ and the "runtime API" which is higher level and used for example in hipSYCL. The concept of active device is a "runtime API" concept and it's not in the "driver API", but it would make sense to touch on that in this CUDA backend spec as well.

I had another look through the CUDA Runtime and Driver APIs and as far as I can tell devices and contexts are the only objects to have this concept of being "active".

I'll have a second look at this and the `device` description to try and clarify this a bit and maybe make it more friendly to the runtime API as well.